### DISCUSSION TITLE
External models with no behavior input/output still requires a Behavior definition

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#104] [Archive] [answered]
External models with no behavior input/output still requires a Behavior definition. While this is technically valid, it looks weird to define an empty behavior for a model with no inputs/outputs.

Should we remove behavior from the required model fields?

_Originally posted by @Coffee2Bits in https://github.com/jondavid-black/AaC/pull/102#discussion_r746133539

**Comment by Coffee2Bits** [comment]
This also raises some ancillary problems.

1. How do we handle empty array fields? Do we do what @cvchaparro suggests in the PR, or allow the user to omit the field completely?
2. How do we validate a required field that's an array, but is empty?

**Reply by jondavid-black** [reply]
I think "required" should mean a value is provided, not the field is simply present.  For example, if a non-array field was required but NULL I don't think it would make sense.  For an array, I think there should be at least 1 value.

**Reply by Coffee2Bits** [reply]
I agree. Required for field arrays should be at least one.

That being said, if we force that, then we may want to remove `behavior` from the required field for models unless we want all external models to have at least one modeled behavior.

**Reply by cvchaparro** [reply]
I agree that a non-list field that's required might not make much sense if it was `NULL` but I think this is not an accurate comparison. An empty list is still a list just like an empty string is still a string, 0 is still a number, etc. I could see empty strings/empty lists being an anti-pattern (at least in some cases) but I don't think the validator is really where we should be checking for anti-patterns. At least not as it is, currently - validator plugins could change my mind about that.

Using software development as an example, the way I see it, the validator is similar to a pre-processor/compiler and getting a warning (I think this makes most sense as a warning - i.e. technically valid but probably not what you meant) for empty lists of required fields would fall into the realm of a static analysis tool like what SonarQube might do.

**Reply by cvchaparro** [reply]
@Coffee2Bits

Per our discussion earlier, did you have any more comments on this?

**Comment by jondavid-black** [comment]
I don't know if it's related, but I ran into something unexpected yesterday.  I was able to successfully run validate (and get a valid response) when the model actually had a typo in a field name.  Are these observations indicating we have a bug in validation?

**Reply by Coffee2Bits** [reply]
There is currently a bug where the types referenced in usecases aren't being validated. I'm spending some time going through the validator again to identify any other overlooked issues, and I'll document/post my findings and any solutions I come up with.

**Reply by cvchaparro** [reply]
Without more context on what the issue was, it would be challenging to say for sure whether it was a bug or a "feature". That said, I think we've determined that the validator has some edge-cases we need to account for.

**Comment by Coffee2Bits** [comment, marked as answer]
Addressed by issue #99.

**Comment by lizzcondrey** [comment]
This has been marked resolved, has an answer, and has the issue and merge linked. Closing this discussion.
